bentoml/_internal/frameworks/common/pytorch.py

```diff
@@ -41,141 +41,156 @@
 
-    def _shutdown(self) -> None:
-        if self._no_grad_context is not None:
-            self._no_grad_context.close()
-            self._no_grad_context = None
+    def __del__(self):
+        self._no_grad_context.close()
+        # no need for now because our worker process will quit and return the gpu memory
+        # if self.device_id == "cuda":
+        # torch.cuda.empty_cache()
 
-    def _run_batch(
-        self,
-        *args: t.Union["ext.NpNDArray", torch.Tensor],
-        **kwargs: t.Union["ext.NpNDArray", torch.Tensor],
-    ) -> torch.Tensor:
 
+def make_pytorch_runnable_method(method_name: str) -> t.Callable[..., torch.Tensor]:
+    def _run(
+        self: PytorchModelRunnable,
+        *args: ext.NpNDArray | torch.Tensor,
+        **kwargs: ext.NpNDArray | torch.Tensor,
+    ) -> torch.Tensor:
         params = Params[t.Union["ext.NpNDArray", torch.Tensor]](*args, **kwargs)
 
         def _mapping(item: t.Union["ext.NpNDArray", torch.Tensor]) -> torch.Tensor:
             if LazyType["ext.NpNDArray"]("numpy.ndarray").isinstance(item):
-                item = torch.Tensor(item, device=self._device_id)
+                return torch.Tensor(item, device=self.device_id)
             else:
-                item = item.to(self._device_id)
-            return item
+                return item.to(self.device_id)  # type: ignore
 
         params = params.map(_mapping)
-        res = self._predict_fn(*params.args, **kwargs)
-        return res
+        return getattr(self.model, method_name)(*params.args, **params.kwargs)
+
+    return _run
 
 
 class PyTorchTensorContainer(DataContainer[torch.Tensor, torch.Tensor]):
     @classmethod
-    def singles_to_batch(cls, singles, batch_axis=0):
-        return torch.stack(singles, dim=batch_axis)
+    def batches_to_batch(
+        cls,
+        batches: tuple[torch.Tensor] | list[torch.Tensor],
+        batch_dim: int = 0,
+    ) -> t.Tuple[torch.Tensor, list[int]]:
+        # numpy.concatenate may consume lots of memory, need optimization later
+        batch = torch.cat(  # type: ignore[reportGeneralTypeIssues]
+            batches,
+            dim=batch_dim,
+        )
+        indices = list(
+            itertools.accumulate(subbatch.shape[batch_dim] for subbatch in batches)
+        )
+        indices = [0] + indices
+        return batch, indices
 
     @classmethod
-    def batch_to_singles(cls, batch, batch_axis=0):
-        return [
-            torch.squeeze(tensor, dim=batch_axis)
-            for tensor in torch.split(batch, 1, dim=batch_axis)
-        ]
+    def batch_to_batches(
+        cls,
+        batch: torch.Tensor,
+        indices: t.Sequence[int],
+        batch_dim: int = 0,
+    ) -> t.List[torch.Tensor]:
+        sizes = [indices[i] - indices[i - 1] for i in range(1, len(indices))]
+        return torch.split(batch, sizes, axis=batch_dim)
 
     @classmethod
     @inject
-    def single_to_payload(
+    def to_payload(  # pylint: disable=arguments-differ
```

Comment:
`to_payload` should take a `batch_dim` now and set payload length for validation on the service app side.

Reply:
done